core/src/main/scala/cats/data/NonEmptyList.scala

```diff
@@ -165,6 +165,26 @@ final case class NonEmptyList[A](head: A, tail: List[A]) {
     go(head, tail, Nil)
   }
 
+  /**
+   * Zips each element of this `NonEmptyList` with its index.
+   *
+   * {{{
+   * scala> import cats.data.NonEmptyList
+   * scala> val nel = NonEmptyList.of("a", "b", "c")
+   * scala> nel.zipWithIndex
+   * res0: cats.data.NonEmptyList[(String, Int)] = NonEmptyList((a,0), (b,1), (c,2))
+   * }}}
+   */
+  def zipWithIndex: NonEmptyList[(A, Int)] = {
+    @tailrec
+    def go(rest: List[A], ix: Int, acc: List[(A, Int)]): List[(A, Int)] =
```

Comment:
Way to be rigorous! So, if I'm reading this right, if the `tail != Nil` (listSize > 1) this is ~ 2-3x faster to use the builder. I would have assumed 2x or so because you skip the reverse.